Uncalled function checks only works with single conditional

jonhue opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

This pull request makes two changes:

1. it checks all operands of `||` binary expressions for uncalled functions (not just the rightmost ones)
2. it checks operands of `!` unary expressions for uncalled functions (if the function cannot be `undefined`)

Fixes #35584

jonhue commented on src/compiler/checker.ts:
```diff
@@ -34560,12 +34564,17 @@ namespace ts {
             }
 
+            const testedType = location === condExpr ? type : checkExpression(location);
+            if (typeToString(testedType).includes('undefined')) { // maybeTypeOfKind(testedType, TypeFlags.Undefined)
```
I don't really understand why, but `maybeTypeOfKind(testedType, TypeFlags.Undefined)` always returns `false`, even for types like `() => boolean | undefined`. This is because for some reason that I am unaware of the type ``() => boolean | undefined`` is lacking the `Union` flag. So what I put into the `if` here is just a placeholder.

jonhue commented on tests/baselines/reference/truthinessCallExpressionCoercion2.errors.txt:
```diff
@@ -87,8 +86,6 @@ tests/cases/compiler/truthinessCallExpressionCoercion2.ts(112,14): error TS2774:
         // error
         typeof window !== 'undefined' && window.console &&
             ((window.console as any).firebug || (window.console.exception && window.console.table));
-                                                 ~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?
```
This error disappeared as a consequence of the temporary fix I mentioned in https://github.com/microsoft/TypeScript/pull/42835/files#r577513285. It will appear again once that is resolved.

sandersn commented:
@typescript-bot user test this
@typescript-bot run dt

@jonhue A couple of things to do next:

1. Look to see if there are new errors from the user tests or DT tests. Report on whether the new errors are useful or not. (The DT tests may have lots of false positives, since they're often not real code.)
2. Open a separate bug for "uncalled function checks don't work with negation".

jonhue commented:
@sandersn Thank you for your comment! I skimmed through the test suite run and the errors look reasonable to me.
I also split this pr up into two, as you suggested. This one, and #43097. 🙂

sandersn commented on src/compiler/checker.ts:
```diff
@@ -34550,6 +34550,9 @@ namespace ts {
             }
 
             const location = isBinaryExpression(condExpr) ? condExpr.right : condExpr;
+            if (isBinaryExpression(condExpr) && condExpr.operatorToken.kind === SyntaxKind.BarBarToken) {
```
should this work on `&&` too? It seems like it but I haven't thought about it too deeply.
Edit: From the issue it sounds like we want to enable one operator at a time to make sure we don't introduce lots of false positives.

jonhue replied:
I agree. I think enabling this for `&&` too could also catch many bugs like in 

```ts
declare function isFoo(): boolean;
if (isFoo && somethingElse) {}
```

The added difficulty will be handling values that might be `undefined`. In the case of 

```ts
declare const isFoo: (() => boolean) | undefined;
if (isFoo && somethingElse) {}
```

we would not want to throw an error. This is a similar complexity to what we have with negations (#43096). We don't have this problem with the `||` operator because if the value is `undefined` we would just move on to the right operand anyway.

An additional problem might be with incomplete type definitions of libraries where functions lack the `undefined` type. I'm not sure how big of a problem this might turn out to be and I don't think we should let this inform too much of what we do here but it is something to keep in mind.

jonhue replied:
We do already check conjunctions but there is a related case where we do not check values that are referenced in the block. I opened #43211 to maybe do some experimentation on that front.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -34550,6 +34550,9 @@ namespace ts {
 
             const location = isBinaryExpression(condExpr) ? condExpr.right : condExpr;
+            if (isBinaryExpression(condExpr) && condExpr.operatorToken.kind === SyntaxKind.BarBarToken) {
+                checkTestingKnownTruthyCallableOrAwaitableType(condExpr.left, type, body);
```
I don't think recursion is a good idea here, because binary expressions might be pretty long, and because the core of `checkTestingKnownTruthyCallableOrAwaitableType` is not too complex.

I'd move the `getSignatureOfType` check before `location`'s declaration, then put the rest of the function in a helper that's called in a loop, something like:

```ts
            helper(condExpr, body)
            while (isBinaryExpression(condExpr) && condExpr.operatorToken.kind === SyntaxKind.BarBarToken) {
                condExpr = condExpr.left
                helper(condExpr, body)
            }
```

(although you might be able to write it more elegantly)

sandersn commented on tests/baselines/reference/uncalledFunctionChecksInConditional.errors.txt:
```diff
@@ -0,0 +1,34 @@
+        // error
+    }
+    
+    if (isFoo || isBar) {
```
just for safety, can you add a test like `x || y || z() || foo`?

sandersn commented:
Actually, looking at the whole call chain, I think `type` needs to be checked per-`condExpr` as well. At least there needs to be tests for multiple, alternating defined/undefined uncalled functions, like `x || uy || z || ufoo` or `ux || y || uz || foo`, where `u-` means that the variable has undefined in its type.

jonhue commented:
@sandersn Thank you for your comments! You're right, we do need to check the type per `condExpr` for this to work. I added some more test cases and it does work as we would expect now 🙂.
I also implemented the refactoring you suggested.

sandersn commented:
I like the code as-is, but it checks a lot more than before. I want to see whether there are many new errors in the user tests:

@typescript-bot user test this
@typescript-bot run dt

@jonhue can you take a look at those after they're done running?

jonhue commented:
@sandersn looks like something went wrong while running the tests. I can't really gather from the logs what exactly went wrong. Could yo take a brief look/rerun the tests? 🙂

sandersn commented:
@typescript-bot user test this
@typescript-bot run dt

sandersn commented:
No report from typescript-bot again =( I manually checked the DT run and there weren't any new failures. It looks like the user test run failed to post a comment after it successfully pushed to the existing PR at https://github.com/typescript-bot/TypeScript/pull/158/files

I looked at the existing PR and saw two new errors: [the fluentui one](https://github.com/microsoft/fluentui/blob/5254c862ebd2ac5da1891cc1c1271097e23bd01b/packages/react-experiments/src/components/FloatingSuggestions/FloatingSuggestions.tsx#L281) is good. It looks like the author forgot that `hasSuggestionSelected` was a method since it's used right before an accessor. [The clone error](https://github.com/pvorb/clone/blob/6df949a1f2175763319326a4e1c4911a9acbd878/clone.js#L263) is on `module.exports` -- TS generally doesn't understand module checks, so it's not a problem.

sandersn commented:
Since this is a new feature, I'll merge this after the 4.3 beta is done.

clarkio commented:
Thank you all for your work in this 👍 Glad to see it added
